### DISCUSSION TITLE
v6.4: How to show Pending UI in case of slow loader

### DISCUSSION THREAD

**Discussion by devpeerapong** [remix-run/react-router#8914] [Q&A] [answered]
Super excite about v6.4 release and now I'm migrating my project to use `loader` and `action`.
But unfortunately our backend on dev env has super slow cold start (~5 seconds). 
So when we click the side menu to navigate. Nothing happens. Even the active state on the menu.
It seems like `react-router` is waiting for the loader to resolve and then start the transition.

Is there a way to show Pending UI when the loader is loading ?
Like when you use `React.lazy` with `Suspense`
I skim through the [Remix ](https://remix.run/docs/en/v1/tutorials/blog) & [React Router 6.4](https://beta.reactrouter.com/en/remixing/getting-started/data#data-loading) Documentation but found nothing about Pending UI.

Here is the demo.
https://stackblitz.com/edit/github-e5anhv?file=src%2Froutes%2Froot.jsx

similar to the demo in v6.4 release, but with the delay on `noteLoader` and `NavLink` on the side navigation.
notice that when we click the note, nothing happens. Until 2 seconds later then the page appears and the link now becomes red
(active state).

**Comment by tommie-lie** [comment]
`<Suspense>` relies on a promise being thrown, which the `<Outlet>` doesn't do while the route's loader is still running.
Instead, you can use [`useNavigation`](https://beta.reactrouter.com/en/dev/hooks/use-navigation) to access the current navigation state and show a loading component:
```javascript
export default function Root() {
  const notes = useLoaderData();
  const { state } = useNavigation();

  return (
    <>
      {state === 'loading' && (
        <progress
          style={{
            position: 'absolute',
            top: '0',
            left: '0',
            width: '100vw',
          }}
        />
      )}
      <div style={{ display: 'flex' }}>
        {/* same code as before */}
      </div>
    </>
  );
}```

**Reply by devpeerapong** [reply]
Thank for the suggestion, I was able to show loading state when page changed and now got hit with another issue.
`useNavigation` only works for [global loading indicators](https://beta.reactrouter.com/en/dev/hooks/use-navigation#usenavigation).
If your nested route has a loader that depends on URLSearchParams (ex. pagination) and you apply this pattern then your parent and nested route will both show loading indicator.

as you can see in https://stackblitz.com/edit/github-e5anhv-w9jwwf

When you switch notes or paginate comments, you will always see both `Page Loading` and `Comment Loading`.
Is there a way to limit the navigation in the route scope ? like `useLoaderData` for route's loader

**Reply by sanelaxm** [reply]
@devpeerapong Any luck finding a work around? I have the same exact scenario I am running into

**Comment by pawelphilipczyk** [comment]
I'm trying to do the same thing so I'll reiterate the question asked. I've got this example that has a home page and about page: 
https://codesandbox.io/s/remix-router-in-react-js-forked-2uuqcd

When I navigate to the about page which has an artificially delayed `loader`, I can show a spinner in a similar way as pointed above. 

The problem is it seems the route does not render my "page" component until its `loader` is finished. In this case, is there a possibility to render a fallback element for it? 

Adding to that also seems the `fallbackElement` on the `DataBrowserRouter` in my example is not rendering as well.

**Reply by tommie-lie** [reply]
Well, that's kinda what the data routing is there for: render the route only when the data is available. What do you expect the rendering to look like before your `loader` is finished after you have clicked the About button? Do you want to *not* show the previous page? In this case, you can render your `<Outlet />` conditionally: https://codesandbox.io/s/remix-router-fallback-ui-forked-kz0x1l?file=/src/App.js

You can also replace the `<Outlet />` with a full-screen gray overlay (to indicate that the user cannot interact with the page) or a full-screen animation.
Generally, it's up to you to decide what should happen during this "in-between" loading state. The default behaviour of react-router mimicks the browser's behaviour in a SSR/static content sitation: show the old page until the new page is loaded. If you want to override this behaviour, I think using the navigation state like above to replace the old page is the only solution.


About your issue with `<DataBrowserRouter>`'s `fallbackElement` prop: I think this is a problem with your Ionic setup. The `fallbackElement` is rendered *instead of all children* during the first mount of `<DataBrowserRouter>` until the first route has completely been resolved. This means you render a `<IonSpinner>` component (your fallback) *before* mounting your `<App>` component. I made your example show the fallback by moving the call to `setupIonicReact` just before the `root.render` call (in index.js), but I don't know anything about Ionic, so I'm not sure whether this is the intended way to use it or if you should rather put your routing into a `<IonApp>`.

P.S.: I noticed you tried passing a `fallbackElement` prop to a `<Route>` element. `<Route>`s don't have a fallback.

**Reply by mattiaz9** [reply]
> Well, that's kinda what the data routing is there for: render the route only when the data is available.

I wish that was a user's choice. Sometimes it makes more sense, or maybe is better for the ux, to show a placeholder loading.

Something like this:
```tsx
// default behaviour
<Route path="/" loader={...} loadingMode="blocking"  />

// not blocking
<Route path="/" loader={...} loadingMode="render"  />
```

And in the component:
```tsx
type PageData = { ... }

const Page = () => {
  const { data, isLoading } = useLoaderData<PageData>()

  if (isLoading) return <Placeholder />

  return <DefaultPage />
}
```

When in blocking mode `isLoading` would be always false because we already have the data.

**Reply by tommie-lie** [reply]
I somehow feel it already is the user's choice. You can opt-out of data routes and roll your own data fetching, using swr, react-fetch-hook, react-query or using useEffect/useState directly. I think those libraries might be closer to what you are expecting if you want to have a `isLoading` state.
Or you can achieve something similar to your example by using the structure from my sandbox and putting the placeholder stuff in a layout route. Sure, it's not a prop on the `<Route>` component, but is it really that common to have this behaviour only on one single route in a larger application? I think most often you'll have some sort of layout route that provides a menu, titlebar, other stuff, and then have a "main area" which differs by route.
Basically like this:
```js
function PageWithLoadingPlaceholder() {
  const { state } = useNavigation();
  
  if (state === "loading") {
    return <Placeholder />;
  }

  return <Outlet />;
}

function App() {
  return (
    <DataBrowserRouter>
      <Route path="/" element={<Layout />}>
        <Route element={<PageWithLoadingPlaceholder />}>
          <Route index loader={loader} element={<Page />} />
          <Route path="/a" loader={aLoader} element={<PageA />} />
          <Route path="/b" loader={bLoader} element={<PageB />} />
        </Route>
      </Route>
    </DataBrowserRouter>
  );
}
```

**Comment by pawelphilipczyk** [comment]
I've played a bit more with those loaders, and the above approach is okay in some cases. 

The problem is when I want to render a different skeleton per route - that was what I originally tried to solve.

Here's what I've come up with so far: 
* Some routes' elements are rendered inside `Suspense` with different `fallback` per element,
* There is a `withSuspense` higher-order function to wrap loaders of those routes,
* Components rendered by routes with suspense have to call the value returned from `useLoaderData` because it's a function,
* This is the first time I'm trying to use Suspense in any real-world scenario so I'm pretty sure things could be improved with Proxy or some suspense-enabled data fetching library would do the trick. 

https://stackblitz.com/edit/github-tyjpmp-n9rngm?file=src%2Fmain.jsx

Note: for some reason when I edit something and save I'm getting this ```A router only accepts one active subscriber``` error and I'm not sure it's something wrong with Stackblitz or my code (probably the latter one is true).

**Reply by pawelphilipczyk** [reply]
Looking at this PR I'm thinking this is exactly what will be handled in a better way with deferred responses: 
https://github.com/remix-run/react-router/pull/9061/files

**Comment by brophdawg11** [comment, marked as answer]
Yep `defer` is what y'all are looking for here - let us know if this solves your use cases :) 

https://beta.reactrouter.com/en/dev/guides/deferred

Also FWIW, the other idiomatic approach (before `defer`) would have been to `useFetcher` in your components and show skeletons/data based on those calls which would give you the "render then fetch" behavior you're looking for, instead of the default "fetch then render" behavior.

**Reply by devpeerapong** [reply]
Hi again.

After playing with some `deferred`. It's not quite what I was looking for.

What I am looking for is to display Pending UI
1. when navigate into the route (ex: `/invoices/1`)
2. when navigate between the route (`/invoices/1` -> `/invoices/2`)

The `<Awaited />` only displays Pending UI once at page loaded.
But when navigate from `/invoices/1` to `/invoices/2`.
The Pending UI in `<Suspense>` will not display again.

I solve this by using `useNavigation` and `matchPath` to check if we should display the Pending UI.
Is there a way to use only `deferred` ?

Demo:
https://stackblitz.com/edit/vitejs-vite-nmhpvx?file=src%2Fpages%2FInvoice.tsx

Snippet from demo:
```jsx
const navigation = useNavigation();
const isSubmitting = navigation.state === 'submitting';
const isNavigationOnCurrentRoute = matchPath(
  '/sales/invoices/:invoiceId',
  navigation.location?.pathname ?? ''
);

const isLoading = navigation.state === 'loading' && isNavigationOnCurrentRoute;

return (
  <Suspense fallback={<Loading>Load Invoice Detail ...</Loading>}> // <-- Here
      <Await resolve={invoice} errorElement={<div>Error Loading</div>}>
        {(data: LoaderData) => {
          if (isLoading) {
            return <Loading>Load Invoice Detail ...</Loading>; // <-- and Here
          }

          return (
            <div className="p-12">
```

**Reply by brophdawg11** [reply]
Good catch @devpeerapong - and thank you for the reproduction!  This is not the expected behavior and we should be showing the pending UI on those navigations from `/invoices/1 -> /invoices/2`.  This is fixed in https://github.com/remix-run/react-router/pull/9181 👍

**Reply by JakubAndrysek** [reply]
I will also mention this example with a Deferred Fetch explanation: 
https://www.infoxicator.com/react-router-6-deferred-fetch

**Reply by maxine-l-m** [reply]
I personally think it would be much nicer if you could just specify a fallback component on the RouteObject instead of needing to use deferred + Await + Suspense, adds a lot of boilerplate to something that could be a small configuration change

**Reply by danillewin** [reply]
@brophdawg11 Hello!

I've found an issue, when navigating via SearchParams instead of PathParams, for example during the pagination:

`/invoices?page=1` -> `/invoices?page=2`

I've described my guess [at your PR](https://github.com/remix-run/react-router/pull/9181#issuecomment-1529135599) and hope for your help)

Thanks in advance)

**Reply by brophdawg11** [reply]
This is expected, since by design we do not fallback on revalidations _within the same route_.  `/invoices?page=1 -> /invoices?page=2` are still the same `/invoices` route.  In this case I would be using `useNavigation` to toggle your loading indicator.

**Reply by danillewin** [reply]
Ok, seems strange at first sight, thank you for your answer

**Comment by hjonasson** [comment]
Before filing my own ticket, I followed this pattern of `defer` in the `loader` + `<Suspense />` and `<Await />`. However, I have a blank screen until all (4) loaders have resolved. One difference is that I am on React 17.0.1 which I don't have control over. Will that make a difference 😕 and upgrading to 18 will give me access to the beautiful loading pattern? 

I tried using `useNavigation` as well in hopes that I would get `state === 'loading` would be true, but on my blank screen nothing is rendered so of course I don't 🤦 .

**Reply by brophdawg11** [reply]
`<Suspense>` is a React 18 feature so you will need to upgrade before you can leverage `defer`/`<Await>`.  Until then you could fetch your slow data via `fetcher.load` calls.

**Reply by franklinjavier** [reply]
> React 16.6 added a <Suspense> component that lets you “wait” for some code to load and declaratively specify a loading state

https://17.reactjs.org/docs/concurrent-mode-suspense.html

**Reply by einarq** [reply]
> `<Suspense>` is a React 18 feature so you will need to upgrade before you can leverage `defer`/`<Await>`. Until then you could fetch your slow data via `fetcher.load` calls.

Do you mean Suspense for data loading? Suspense itself has been available for a long time, but only did something meaningful when used with React.lazy.

**Reply by einarq** [reply]
@brophdawg11 
I'm struggling with showing loading states when navigating in combination with loaders. Like someone else rasied here, I want to show immediate feedback when navigating.
https://github.com/remix-run/react-router/discussions/10540

Seems the recommendation is sometimes to use "defer", but its almost surprising that one would ever do this *without* using defer, which also makes me feel that I'm probably doing something wrong.
A loader can take a long time to load due to network issues (not just slow db calls), so the UI cannot just be hanging until it resolves.
So I went with defer for now.

But then I want to improve the error handling in my app to also use the errorElement on a more global level:
https://reactrouter.com/en/main/hooks/use-route-error

Is it correct that using defer doesnt work out of the box when throwing a Response like this? Do I need to use some other api when doing that?

Or is it better to rely on useNavigation hook for this type of loader, since I then dont need to use defer?

**Reply by brophdawg11** [reply]
It's a product/UX decision at the end of the day as to which to use.  Do you want to keep showing the user the current content during the navigation or do you want to flip to an empty page with spinners/skeletons.  I personally prefer the former so as not to yank valid content out of the UI without valid content to replace it with.  Plus skeletons/spinners are _very_ hard to get right and not introduce CLS/jank.

* without `defer` - keep the user on the current page during loading, showing _valid_ content, and potentially a spinner via `useNavigation`
* with `defer` - navigate away from the current page ASAP, and show skeletons/spinners on the (potentially empty) destination page while data loads

> the UI cannot just be hanging until it resolves

The UI is not hanging while loaders resolve.  It's still a responsive/interactive react app.

> Is it correct that using defer doesn't work out of the box when throwing a Response like this? Do I need to use some other api when doing that?

Not totally - you can throw a `Response` before returning `defer` and it'll bubble to the error boundary. But once you're returned the `defer` the Response is already sent, so you can't then throw _another_ response.  You should just reject the deferred promise with the Error or message you want to display to the user and either (1) access it via `useAsyncError` (https://remix.run/docs/en/main/hooks/use-async-error) if you want to show a contextual error message, or (2) access it with `useRouteError` at the route error boundary if you do not have an `<Await errorElement>`

**Reply by einarq** [reply]
Ok, so where I was probably tripping up was that I was throwing the response, instead of rejecting the promise I pass back in the defer.

Maybe worth mentioning in the docs?

**Reply by einarq** [reply]
> The UI is not hanging while loaders resolve. It's still a responsive/interactive react app.

As someone else pointed out, the app has type display some type of indication that something is happening. What you have in the Remix docs looks good, but having it on a more granular level seems to involve a lot of manual checking of to and from for locations.
Unless I'm misunderstanding something (the navigation state is global, so it cannot easily tell me if the navigation is on a sub-route level).

**Reply by brophdawg11** [reply]
Would you want to open a PR to update the docs with some clarification?

Yep - currently if you want contextual loading indicators you would need to check again the `useNavigation().location`.  In the future you'll be able to leverage `useTransition` and the exposed routing promises (<https://github.com/orgs/remix-run/projects/5?pane=issue&itemId=62177552>) now that React has introduced some first-class APIs for dealing with async operations.

**Reply by einarq** [reply]
Btw, is it intended behavior that the useNavigation state doesn't change if a route doesn't have a loader?

When using the useNavgation hook to control the state of a Top Loader component, I noticed that any route that doesn't have a loader doesn't affect that state.

```
      children: [
        {
          id: CLIENT_ROUTE_NAMES.INDEX,
          index: true,
          element: <RouteHandlers.SurveyList />,
          ErrorBoundary: RouteBoundary,
        },

const navigation = useNavigation(); // doesn't change
```


```
      children: [
        {
          id: CLIENT_ROUTE_NAMES.INDEX,
          index: true,
          element: <RouteHandlers.SurveyList />,
          ErrorBoundary: RouteBoundary,
          loader() {
            return null;
          },
        },

const navigation = useNavigation(); // now it changes
```

**Reply by brophdawg11** [reply]
Yes - if the navigation can be done 100% synchronous then there is a never a pending "navigating" state to render so `useNavigation` never goes into a `loading` state.

Said another way, if you don't have a `loader` to run, and you already have the new route `Component` loaded in the browser - you probably don't want to show a spinner and would prefer to just show the new UI immediately.

**Reply by einarq** [reply]
Ok. Just curious. If I was using the "lazy" feature on a route here, also without a loader (data loading is still done by some of the components, not all from loaders yet), would it then go into loading state?

**Reply by einarq** [reply]
And another things:
Any suggestions on how to figure out if a route is within a certain parent or not?

Is matchPath the best approach, or would you suggest something else? Any examples you could point me to?

**Reply by brophdawg11** [reply]
May I suggest moving these types of questions into the [Remix Discord](https://rmx.as/discord)?  You'll be able to get more help from a large community of users in there :)

**Reply by einarq** [reply]
> May I suggest moving these types of questions into the [Remix Discord](https://rmx.as/discord)? You'll be able to get more help from a large community of users in there :)

I'll try to use Discord more, but I've found it a bit difficult to use for other projects. Difficult to follow a specific thread, everything gets a bit jumbled together. Skill issue for sure :)

In any case, I cannot get deferred loading to work with central error handling, not sure what I'm doing wrong.

I want: 
- a central error component in the router (using ErrorBoundary)
- in that component, I want to use isRouteErrorResponse so that I can catch 404 there (for instance)
- my loader looks something like what I have below, but isRouteErrorResponse is false.

Does isRouteErrorResponse not work when using defer+reject?

```
function loader() {
  const promise = new Promise(async (resolve, reject) => {
    try {
      const data = await myApi();

      resolve(data);
    } catch (e) {
      if (e.status === 404) {
        reject(<what to reject with here, which will be compatible with isRouteErrorResponse >);
      } else {
        reject(e);
      }
    }
  });

  return defer({promise});
}
```

**Comment by vincerubinetti** [comment]
A little bit tangential to what others are discussing, but since this is the first google result for "react router data loader show progress"...

I can't use suspense because I'm doing an old fashioned vite SPA, and not using a "suspense-enabled data source". As such, I did a very simple and kind of dumb solution:

```tsx
// global loading state (jotai)
export const loading = atom(0);

// root router layout
const Layout = () => {
  const { state } = useNavigation();
  const getLoading = useAtomValue(loading);

  return (
    <>
      <>
        <Outlet />
        {/* floating progress bar at top of screen */}
        {state === "loading" && <ProgressBar progress={getLoading} />}
      </>
    </>
  );
};

// some route data loader
async function somePageLoader() {
  getDefaultStore().set(loading, 0.1);
  await someAsyncFunc();
  getDefaultStore().set(loading, 0.2);
  await someAsyncFunc();
  // etc...
}
```

If anyone knows a less naive way to do this, let me know. I guess it would be cool if a `LoaderFunction` could be passed a `progress` function along with `params` and the like, that you could call to update a loading progress % which you could then get from `useNavigation`.

**Reply by slivingston-isn** [reply]
This is the conclusion we arrived at. Using data loaders requires an app-level progress bar or spinner of some kind.

**Comment by einarq** [comment]
Any suggestions on how to figure out if a route is within a certain parent or not?Is matchPath the best approach, or would you suggest something else? Any examples you could point me to?Thanks On 11 Jun 2024, at 16:16, Matt Brophy ***@***.***> wrote:﻿
Would you want to open a PR to update the docs with some clarification?
Yep - currently if you want contextual loading indicators you would need to check again the useNavigation().location.  In the future you'll be able to leverage useTransition and the exposed routing promises (https://github.com/orgs/remix-run/projects/5?pane=issue&itemId=62177552) now that React has introduced some first-class APIs for dealing with async operations.

**Comment by mustafazeydani** [comment]
If you are using RouterProvider, you can specify the fallbackElement directly. This way, you don't need to use the defer function, as the fallbackElement will be displayed until the loader function resolves the promise.
